File: file/src/main/scala/farjs/file/FilePlugin.scala
package farjs.file

import farjs.filelist._
import farjs.filelist.stack._
import scommons.react.ReactClass

import scala.scalajs.js

object FilePlugin extends FileListPlugin(js.Array("M-v")) {

  override def onKeyTrigger(key: String,
                            stacks: WithStacksProps,
                            data: js.UndefOr[js.Dynamic] = js.undefined): js.Promise[js.UndefOr[ReactClass]] = {

    js.Promise.resolve[js.UndefOr[ReactClass]](createUi(key).map(FilePluginUi(_)))
  }
  
  private[file] final def createUi(key: String): js.UndefOr[FilePluginUiParams] = {
    key match {
      case "M-v" => FilePluginUiParams(showFileViewHistoryPopup = true)
      case _ => js.undefined
    }
  }
}

File: file/FilePlugin.mjs
/**
 * @typedef {import("@farjs/filelist/FileListData.mjs").ReactComponent} ReactComponent
 * @typedef {import("./FilePluginUi.mjs").FilePluginUiParams} FilePluginUiParams
 */
import FileListPlugin from "@farjs/filelist/FileListPlugin.mjs";
import FilePluginUi from "./FilePluginUi.mjs";

class FilePluginImpl extends FileListPlugin {
  constructor() {
    super(["M-v"]);
  }

  /**
   * @param {string} key
   * @returns {Promise<ReactComponent | undefined>}
   */
  onKeyTrigger(key) {
    const uiParams = this._createUiParams(key);
    const res = uiParams ? FilePluginUi(uiParams) : undefined;
    return Promise.resolve(res);
  }

  /**
   * @param {string} key
   * @returns {FilePluginUiParams | undefined}
   */
  _createUiParams(key) {
    if (key === "M-v") {
      return {
        showFileViewHistoryPopup: true,
      };
    }

    return undefined;
  }
}

const FilePlugin = new FilePluginImpl();

export default FilePlugin;
